src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -537,6 +541,132 @@ private bool BangTag()
+                {
+                    lastDoubleHyphen = s;
+                    return true;
+                }
+
+                return false;
+            });
+
+            NextToken();
+            EnsureCurrent();
+
+            if (At(HtmlSymbolType.Text) && IsDashSymbol(CurrentSymbol))
+            {
+                // Doing this here to maintain the order of symbols
+                if (!NextIs(HtmlSymbolType.CloseAngle))
+                {
+                    Accept(lastDoubleHyphen);
+                    lastDoubleHyphen = null;
+                }
+
+                AcceptAndMoveNext();
+            }
+
+            return lastDoubleHyphen;
+        }
+
+        internal static bool IsDashSymbol(HtmlSymbol symbol)
+        {
+            return string.Equals(symbol.Content, "-", StringComparison.Ordinal);
+        }
+
+        protected bool IsHtmlCommentAhead()
+        {
+            /*
+             * From HTML5 Specification, available at http://www.w3.org/TR/html52/syntax.html#comments
+             * 
+             * Comments must have the following format:
+             * 1. The string "<!--"
+             * 2. Optionally, text, with the additional restriction that the text
+             *      2.1 must not start with the string ">" nor start with the string "->"
+             *      2.2 nor contain the strings "<!--", "-->", or "--!>"
+             *      2.3 nor end with the string "<!-".
+             * 3. The string "-->"
+             * 
+             * */
+
+            if (CurrentSymbol.Type != HtmlSymbolType.DoubleHyphen)
+            {
+                return false;
+            }
+
+            // Check condition 2.1
+            if (NextIs(HtmlSymbolType.CloseAngle) || NextIs(next => IsDashSymbol(next) && NextIs(HtmlSymbolType.CloseAngle)))
+            {
+                return false;
+            }
+
+            // Check condition 2.2
+            var isValidComment = false;
+            LookaheadUntil((s, p) =>
+            {
+                if (s.Type == HtmlSymbolType.DoubleHyphen)
+                {
+                    if (NextIs(HtmlSymbolType.CloseAngle))
+                    {
+                        // Check condition 2.3: We're at the end of a comment. Check to make sure the text ending is allowed.
+                        isValidComment = !IsCommentContentDisallowed(p);
+                        return true;
+                    }
+                    else if (NextIs(ns => IsDashSymbol(ns) && NextIs(HtmlSymbolType.CloseAngle)))
+                    {
+                        // This is also a valid closing comment case, as the dashes lookup is treated with DoubleHyphen symbols first.
+                        isValidComment = true;
+                        return true;
+                    }
+                    else if (NextIs(ns => ns.Type == HtmlSymbolType.Bang && NextIs(HtmlSymbolType.CloseAngle)))
+                    {
+                        isValidComment = false;
```

Comment:
Looks like this is checking if `--!>` is present in the content. Shouldn't this check be part of `isCommentContentDisallowed()`. That way all the checks about the comment content stay within this method.

Reply:
I see that the naming is not ideal here and it causes confusion. I'm going to rename the "IsCommentContentDisallowed" to something better to indicate that it checks only for invalid ending for comment content.

Reply:
IsCommentTerminatorValid ?